lib/src/main/scala/spinal/lib/MasterSlave.scala

```diff
@@ -2,87 +2,199 @@ package spinal.lib
+    *
+    * @deprecated This method can be overriden but not called. Calling this
+    * method is not correct. Call `setAsSlave()` or `intoSlave()` instead.
+    *
+    * This method is named `asXxx` but it does not return `Xxx`.
+    *
+    * This method does not update `isMasterInterface` and `isSlaveInterface`.
+    */
+  def asSlave(): Unit = intoMaster().asInstanceOf[Data].flip()
+}
 
 trait MSFactory {
-  def postApply(interface : IMasterSlave) : Unit = {}
+  // TODO what is this?
+  def postApply(interface: IMasterSlave): Unit = {}
 }
 
-trait MS{
-  def apply[T <: IMasterSlave](i: T) : T
-  def apply(a: IMasterSlave,b: IMasterSlave,c: IMasterSlave*) : Unit = {
-    apply(a)
-    apply(b)
-    for(e <- c) apply(e)
-  }
+/** Declare a port as `master` or `slave`
+  *
+  * There are 4 available syntaxes, which are all equivalent:
+  *
+  * {{{
+  * val braces = master(Flow(Bool))
+  *
+  * val short = master Flow (Bool)
+  *
+  * val spaceful = master port Flow(Bool)
+  *
+  * val variadic = Flow(Bool)
+  * master(variadic)
+  * }}}
+  *
+  * The "braces" syntax is short and generic, but it uses braces.
+  *
+  * The "short" syntax is short, but it is formatted with a space between the
+  * type and its parameters, and it can be used only with:
+  *
+  *   - `Flow`
+  *   - `Stream`
+  *   - `Event`
+  *
+  * The "spaceful" syntax is generic and beatiful, but more verbose.
+  *
+  * The "variadic" syntax can be used with any number of interfaces, but can
+  * be used only if the interfaces are already declared.
+  *
+  * @see [[master]] [[slave]]
+  */
+sealed trait MS {
+
+  /** Override it to define how to apply port specification on a non-null IMasterSlave */
+  protected def applyIt[T <: IMasterSlave](i: T): T
+
+  /** Declare a port without braces, long syntax
+    *
+    * {{{
+    * val input = slave port Flow(Bool)
+    * }}}
+    *
+    * See [[MS]] for other syntax.
+    */
+  def port[T <: IMasterSlave](i: T): T =
+    if (i != null) applyIt(i)
+    else i
+
+  /** Declare a port with braces
+    *
+    * {{{
+    * val input = slave(Flow(Bool))
+    * }}}
+    *
+    * or
+    *
+    * {{{
+    * val input = Flow(Bool)
+    * slave(input)
+    * }}}
+    *
+    * See [[IODirection]] for other syntaxes.
+    */
+  def apply[T <: IMasterSlave](i: T): T = port(i)
+
+  // TODO: no? def apply[T <: IMasterSlave](data: HardType[T]): T = port(data())
 
+  /** Declare existing interfaces as ports, "variadic" syntax
+    *
+    * {{{
+    * val a = Flow(Bool)
+    * val b = Stream(Bits(5 bits))
+    * master(a, b)
+    * }}}
+    */
+  def apply(is: IMasterSlave*): Unit = is.foreach(port(_))
+
+  // TODO: this is really differen than IODirection
+  // Maybe we can compare the two means and choose the best,
+  // or document why it is not the same in comments.
   object Flow extends FlowFactory {
-    override def postApply(interface : IMasterSlave) : Unit = {
+    // TODO: can't it just be replaced by port(this)
```

Comment:
What would be remplaced ? postApply ?

Reply:
Yes, I have grepped it and it is overridden only here so these `super.postApply` seem to be a no-op.

It seems really related to `Fragment`, I'm not sure I understand the mechanism behind it. It seems like a complex mechanism to apply the same master/slave-ness to another interface; which could be more simple to implement right in `IMasterSlave`?

Reply:
> Yes, I have grepped it and it is overridden only here so these super.postApply seem to be a no-op

Yes they are no-op. The super.postApply(interface) was placed there in case of a day it would not be a no-op anymore.

The usage of postApply is that : 
For a given hardware signal factory (ex StreamFactory), being able to apply some procedural modifier via the execution of postApply (ex setting a stream as slave when called in the context of slave Stream(xxx)

So, should not be in IMasterSlave.

Reply:
Thanks I'll document this a little bit